server/src/main/java/io/crate/replication/logical/ShardTranslogSequencer.java

```diff
@@ -0,0 +1,109 @@
+ * license agreements.  See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.  Crate licenses
+ * this file to you under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial agreement.
+ */
+
+package io.crate.replication.logical;
+
+import io.crate.replication.logical.action.ReplayChangesAction;
+import io.crate.replication.logical.action.ShardChangesAction;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.action.support.replication.ReplicationResponse;
+import org.elasticsearch.client.Client;
+import org.elasticsearch.common.logging.Loggers;
+import org.elasticsearch.index.shard.ShardId;
+
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * A TranslogSequencer allows multiple producers of [Translog.Operation]s to write them in sequence number order to an
+ * index. Producer can call the [send] method to add a batch of operations to the queue. Operations can be
+ * sent out of order i.e. the operation with sequence number 2 can be sent before the operation with sequence number 1.
+ * In this case the Sequencer will internally buffer the operations that cannot be delivered until the missing in-order
+ * operations arrive.
+ *
+ * Derived from org.opensearch.replication.task.shard.TranslogSequencer
+ */
+public class ShardTranslogSequencer {
+
+    private static final Logger LOGGER = Loggers.getLogger(ShardTranslogSequencer.class);
+
+    private final ShardId shardId;
+    private final Client client;
+    private final long initialSeqNo;
+    private final ConcurrentHashMap<Long, ShardChangesAction.Response> unAppliedChanges = new ConcurrentHashMap<>();
+
+    public ShardTranslogSequencer(ShardId shardId,
+                                  Client client,
+                                  long initialSeqNo) {
+        this.shardId = shardId;
+        this.client = client;
+        this.initialSeqNo = initialSeqNo;
+    }
+
+    public void send(ShardChangesAction.Response response) throws Exception {
+        unAppliedChanges.put(response.fromSeqNo(), response);
+        process();
+    }
+
+    private synchronized void process() throws Exception {
```

Comment:
It appears as if there is only a single producer. Am I missing something, or could we remove the synchronized and use a regular map?

Reply:
Currently yes, OpenSearch uses parallel changes reader per shard (default: 2) and I wasn't sure if we also want to do that? Forgot to mention this for discussion. If not we could probably remove this whole class.

Reply:
As discussed, +1 for removing this or now (but we should add a tracking issue for optimization later)